### PR TITLE
fix: :bug: use unknown module type in SWC

### LINKED ISSUE #2339: [Bug Report]: Parses all JS files in strict mode
### System Info

```
  System:
    OS: Linux 6.2 NixOS 23.05 (Stoat) 23.05 (Stoat)
    CPU: (12) x64 Intel(R) Core(TM) i7-10710U CPU @ 1.10GHz
    Memory: 8.50 GB / 15.27 GB
    Container: Yes
    Shell: 5.2.15 - /run/current-system/sw/bin/bash
  Binaries:
    Node: 18.14.2 - ~/.nix-profile/bin/node
    Yarn: 1.22.19 - ~/.nix-profile/bin/yarn
    npm: 9.5.0 - ~/.nix-profile/bin/npm
  Browsers:
    Chromium: 111.0.5563.64
    Firefox: 111.0b8
  npmPackages:
    @rspack/cli: ^0.1.1 => 0.1.1 
```

### Details

rspack has inherited SWC’s default of parsing all JS files as modules, and therefore cannot parse code that only compiles in non-strict mode, such as

```js
delete a;
```

or as a real-world example, the [sorttable](https://www.npmjs.com/package/sorttable) library.

SWC provides the `isModule` configuration option for this; I think we want to set it based on the file extension (`.cjs`, `.mjs`, `.cts`, `.mts`) or the [`type` field](https://nodejs.org/api/packages.html#type) in the nearest `package.json` (`.js`, `.ts`).

See for example

- swc-project/swc#1810

Probably related:

- #2178

### Reproduce link

_No response_

### Reproduce Steps

From an empty directory:

```console
$ mkdir src
$ echo 'delete a;' > src/index.js
$ rspack
Hash: eb46cf29de5a8584
PublicPath: auto
  Asset       Size  Chunks             Chunk Names
main.js  248 bytes    main  [emitted]  main
Entrypoint main = main.js
[887] ./src/index.js 10 bytes {main}

error[javascript]: JavaScript parsing error
  ┌─ src/index.js:1:8
  │
1 │ delete a;
  │        ^ 'delete' cannot be called on an identifier in strict mode


build: 29.688ms
```

### PR DESCRIPTION
**Pull request by JSerFeng**
## Summary

- Bump SWC

- Should parse CJS as Script instead of Module, we can let SWC automatically do that for us.

## Related issue (if exists)

fixed #2339 

## Types of changes

- [ ] Docs change / Dependency upgrade
- [x] Bug fix
- [ ] New feature / Improvement
- [ ] Refactoring
- [ ] Breaking change

## Checklist

- [x] I have added changeset via `pnpm run changeset`.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: crates/rspack_plugin_javascript/src/visitors/swc_visitor/react.rs [hunk 1]
```diff
@@ -81,40 +81,71 @@ module.hot.accept();
 RefreshRuntime.queueUpdate();
 "#;
 
-static HMR_FOOTER_AST: Lazy<Program> =
-  Lazy::new(|| parse_js_code(HMR_FOOTER.to_string(), &ModuleType::Js).expect("TODO:"));
+static HMR_FOOTER_AST: Lazy<Script> = Lazy::new(|| {
```
Comment (IWANABETHATGUY) [inline crates/rspack_plugin_javascript/src/visitors/swc_visitor/react.rs:84]: Why change `Program` to `Script`?

Comment (JSerFeng) [reply crates/rspack_plugin_javascript/src/visitors/swc_visitor/react.rs:84]: - Script body is Vec<Stmt> while Module body is Vec<ModuleItem> here, Vec<Stmt> can be transformed to Vec<ModuleItem> easily, we need to insert this ast body to both Module and Script

**Review by h-a-n-a — APPROVED** [review]
@hyf0 Can you take a look at the change of tree-shaking part?

**Review by hyf0 — APPROVED** [review]

**Comment by JSerFeng** [comment]
Block by https://github.com/swc-project/swc/issues/7118

**Comment by andersk** [comment]
> Block by https://github.com/swc-project/swc/issues/7118

Fixed in swc ≥ 1.3.45.

**Comment by h-a-n-a** [comment]
@JSerFeng  I haven't dig into the implementation, but could you check this https://github.com/web-infra-dev/rspack/issues/1793? Didn't know if `allowTopLevelThis` is enable by default in type `Unknown`. Related PR: https://github.com/web-infra-dev/rspack/pull/1792

**Comment by JSerFeng** [comment]
> Didn't know if `allowTopLevelThis` is enabled corresponding to type `Unknown` when matched to `script`, I might take manually enabling `allowTopLevelThis` for every module type a mistake. Related PR: #1792

`allow_top_level_this` is placed in `run_after_pass`, we know if a file is `strict` or not at this time, and determine `allow_top_level_this` based on that, I think it should be OK

**Comment by JSerFeng** [comment]
> > Block by [swc-project/swc#7118](https://github.com/swc-project/swc/issues/7118)
> 
> Fixed in swc ≥ 1.3.45.

@andersk  Sorry for the delay, I encountered some more SWC issue in this PR, so it takes this long, now is should be OK to continue

**Comment by JSerFeng** [comment]
@web-infra-dev/rspack please review carefully

**Comment by ahabhgk** [comment]
Some tests snapshot result and webpack.config.js are changed, could you add some explanation?

**Comment by JSerFeng** [comment]
> Some tests snapshot result and webpack.config.js are changed, could you add some explanation?

Stats hash changes because "use strict" disappears when moduleType is asset or script.
Some error snapshots changes because `let` is valid variable name in non-strict mode.

**Comment by JSerFeng** [comment]
Should I extract the bump SWC part to another PR @hyf0

**Review by hyf0 — APPROVED** [review]

**Review by ahabhgk — APPROVED** [review]

**Pull request merged**